airbyte-integrations/postgres-destination/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java

```diff
@@ -0,0 +1,341 @@
+    try {
+      final BasicDataSource connectionPool = getConnectionPool(config);
+      DatabaseHelper.query(connectionPool, ctx -> ctx.execute(
+          "SELECT *\n"
+              + "FROM pg_catalog.pg_tables\n"
+              + "WHERE schemaname != 'pg_catalog' AND schemaname != 'information_schema';"));
+
+      connectionPool.close();
+    } catch (Exception e) {
+      // todo (cgardens) - better error messaging for common cases. e.g. wrong password.
+      return new StandardCheckConnectionOutput().withStatus(Status.FAILURE).withMessage(e.getMessage());
+    }
+
+    return new StandardCheckConnectionOutput().withStatus(Status.SUCCESS);
+  }
+
+  @Override
+  public StandardDiscoverSchemaOutput discover(JsonNode config) {
+    throw new RuntimeException("Not Implemented");
+  }
+
+  /**
+   * Strategy:
+   * <p>
+   * 1. Create a temporary table for each stream
+   * </p>
+   * <p>
+   * 2. Accumulate records in a buffer. One buffer per stream.
+   * </p>
+   * <p>
+   * 3. As records accumulate write them in batch to the database. We set a minimum numbers of records
+   * before writing to avoid wasteful record-wise writes.
+   * </p>
+   * <p>
+   * 4. Once all records have been written to buffer, flush the buffer and write any remaining records
+   * to the database (regardless of how few are left).
+   * </p>
+   * <p>
+   * 5. In a single transaction, delete the target tables if they exist and rename the temp tables to
+   * the final table name.
+   * </p>
+   *
+   * @param config - integration-specific configuration object as json. e.g. { "username": "airbyte",
+   *        "password": "super secure" }
+   * @param schema - schema of the incoming messages.
+   * @return consumer that writes singer messages to the database.
+   * @throws Exception - anything could happen!
+   */
+  @Override
+  public DestinationConsumer<SingerMessage> write(JsonNode config, Schema schema) throws Exception {
+    // connect to db.
+    final BasicDataSource connectionPool = getConnectionPool(config);
+    Map<String, WriteConfig> writeBuffers = new HashMap<>();
+
+    // create tmp tables if not exist
+    for (final Stream stream : schema.getStreams()) {
+      final String tableName = stream.getName();
+      final String tmpTableName = stream.getName() + "_" + Instant.now().toEpochMilli();
+      DatabaseHelper.query(connectionPool, ctx -> ctx.execute(String.format(
+          "CREATE EXTENSION IF NOT EXISTS \"uuid-ossp\";\n"
+              + "CREATE TABLE \"%s\" ( \n"
+              + "\"ab_id\" uuid PRIMARY KEY DEFAULT uuid_generate_v4(),\n"
+              + "\"%s\" jsonb,\n"
+              + "\"ab_inserted_at\" TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP\n"
+              + ");",
+          tmpTableName, COLUMN_NAME)));
+
+      final Path queueRoot = Files.createTempDirectory("queues");
+      final BigQueue writeBuffer = new BigQueue(queueRoot.resolve(stream.getName()), stream.getName());
```

Comment:
How do we handle two sources with the same named streams syncing at the same time?

Reply:
I assumed these would be running in separate workers, but I guess the source shouldn't assume it?

Reply:
different workers. why shouldn't the source assume this? we add noise at the end of the paths to fix this, but i think the contract is that the source runs on its own docker image (with its own filesystem).

Reply:
you're right -- each source should run in its own container. i was being silly in my head .